File: src/gpt_generator/helpers/file_collection.py

import asyncio
import logging
from gpt_generator.modules.worker import gpt_worker
from helpers.architecture_extractor import structure_2_dict
from helpers.code_extractor import extract_code

from models.file_interface import File, FileCollection
from models.gpt_responses_interface import ArchitectorResponse


async def a_create_file_collection(
        project: ArchitectorResponse,
        raw_architector_response: str) -> FileCollection:
  try:
    # Initialize FileCollection
    file_collection = FileCollection(project.name)

    # Add initial response to collection
    file_collection.collections["raw_responses"].append(
        File("architector.txt", str(raw_architector_response))
    )

    structure_dictionary = structure_2_dict(project.structure)

    # Gather all coroutines for gpt_worker_async
    tasks: list[asyncio.Task] = []
    for file in project.files:
      task = asyncio.create_task(
          gpt_worker(project, file.file_name, file.description)
      )
      tasks.append(task)

    responses = await asyncio.gather(*tasks)

    # Add results to the collection
    for file, result in zip(project.files, responses):
      file_collection.collections["raw_responses"].append(
          File(f"raw_{file.file_name}.txt", result, "")
      )
      file_collection.collections["formatted_responses"].append(
          File(file.file_name, extract_code(result),
               structure_dictionary[file.file_name])
      )

    return file_collection

  except Exception as e:
    logging.error(f"An error occurred: {e}")
    return FileCollection("")

File: src/gpt_generator/helpers/file_collection.ts
import gptWorker from '../modules/worker';
import { structure2Dict } from '../../helpers/architecture_extractor';
import { extractCode } from '../../helpers/code_extractor';
import { createFile, createFileCollection, File, FileCollection } from '../../models/file_interface';
import { ArchitectorResponse } from '../../models/gpt_responses_interface';

const createArchitectureFileString = (project: ArchitectorResponse, rawArchitectorResponse: string) => `
## User Prompt ##

## File Structure ##
${project.structure}

## Architecture Description ##
${rawArchitectorResponse}
`;
const addInitialResponse = (fileCollection: FileCollection, architectureFileString: string) =>
  fileCollection.collections['raw_responses']?.push(createFile('architector.txt', architectureFileString, ''));

const processFiles = async (project: ArchitectorResponse, structureDictionary: Record<string, string>) => {
  const tasks = project.files.map(({ fileName, description }) => gptWorker(project, fileName, description));
  const responses = await Promise.all(tasks);
  return project.files.map((file, index) => ({
    rawResponse: createFile(`raw_${file.fileName}.txt`, responses[index] ?? '', ''),
    formattedResponse: createFile(file.fileName, extractCode(responses[index] ?? ''), structureDictionary[file.fileName] ?? ''),
  }));
};

const addResultsToCollection = (fileCollection: FileCollection, results: { rawResponse: File; formattedResponse: File }[]) => {
  results.forEach(({ rawResponse, formattedResponse }) => {
    fileCollection.collections['raw_responses']?.push(rawResponse);
    fileCollection.collections['formatted_responses']?.push(formattedResponse);
  });
};

export const aCreateFileCollection = async (project: ArchitectorResponse, rawArchitectorResponse: string): Promise<FileCollection> => {
  try {
    const fileCollection: FileCollection = createFileCollection(project.name);
    const architectureFileString = createArchitectureFileString(project, rawArchitectorResponse);
    addInitialResponse(fileCollection, architectureFileString);

    const structureDictionary = structure2Dict(project.structure);
    const results = await processFiles(project, structureDictionary);
    addResultsToCollection(fileCollection, results);

    return fileCollection;
  } catch (e) {
    console.error(`An error occurred: ${e}`);
    return createFileCollection('');
  }
};
